petastorm/pytorch.py

```diff
@@ -120,11 +125,39 @@ def __iter__(self):
             # Default collate does not work nicely on namedtuples and treat them as lists
             # Using dict will result in the yielded structures being dicts as well
             row_as_dict = row._asdict()
+
+            # Promote some types that are incompatible with pytorch to be pytorch friendly.
             _sanitize_pytorch_types(row_as_dict)
-            batch.append(row_as_dict)
+
+            # Add rows to shuffling buffer
+            self._shuffling_buffer.add_many([row_as_dict])
+
+            # If the shuffling buffer is ready, pull the resampled rows out and try add to our accumulated batch
+            while self._shuffling_buffer.can_retrieve():
+                post_shuffled_row = self._shuffling_buffer.retrieve()
+                batch.append(post_shuffled_row)
+
+                # Batch is ready? Collate and emmit
+                if len(batch) == self.batch_size:
+                    yield self.collate_fn(batch)
+                    batch = []
+
+        # Once reader can not emit new rows, we might still have a bunch of rows waiting in the shuffling buffer.
+        # Telling shuffling buffer thatt we are finished allows to deplete the buffer completely, regardless its
+        # min_after_dequeue setting.
+        self._shuffling_buffer.finish()
+
+        # Some code duplication with the main reading loop. Not sure how to reorganize to avoid this without
+        # overcomplicating the code too much.
+        while self._shuffling_buffer.can_retrieve():
+            post_shuffled_row = self._shuffling_buffer.retrieve()
+            batch.append(post_shuffled_row)
+
+            # Batch is ready? Collate and emmit
             if len(batch) == self.batch_size:
```

Comment:
nit: I think the final 'if batch` check is fine. But can't we just have this be len(batch) <= self.batch_size. 
Is there any situation where len(batch) would be > self.batch_size here.

Reply:
Not sure I follow. The second `while self._shuffling_buffer.can_retrieve()` loop on line 168, runs down the sample while completely depleting the shuffling queue. 
For example:
- Assume batch size is 9
- In the stead state the shuffling queue has 1000 samples.
- Now we don't have any more samples to load and `for row in self.reader:` exits. We still have 1000 samples awaiting in the shuffling queue.
- We call finish, which enables further depletion of the samples
- We are still able to produce 111 batches of size 9, then pop last sample from the shuffling queue and publish it on line 178.

Did I miss a way to simplify something?

Reply:
Yes, that makes sense. My initial thought that after row in self.reader exists, you can at most have one batch left. That was incorrect. 

Thanks for explaining